sound/soc/sof/sof-audio.c

```diff
@@ -171,6 +171,196 @@ static int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swi
+	case snd_soc_dapm_asrc:
+	case snd_soc_dapm_siggen:
+	case snd_soc_dapm_effect:
+	case snd_soc_dapm_mux:
+	case snd_soc_dapm_demux:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
+					  struct snd_soc_dapm_widget_list *list, int dir)
+{
+	struct snd_soc_dapm_widget *widget;
+	struct snd_soc_dapm_path *p;
+	int ret;
+	int i;
+
+	/*
+	 * Set up connections between widgets in the sink/source paths based on direction.
+	 * Some non-SOF widgets exist in topology either for compatibility or for the
+	 * purpose of conencting a pipeline from a host to a DAI in order to receive the DAPM
+	 * events. But they are not handled by the firmware. So ignore them.
+	 */
+	if (dir == SNDRV_PCM_STREAM_PLAYBACK) {
+		for_each_dapm_widgets(list, i, widget) {
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_sink_path(widget, p)
+				if (p->sink->dobj.private && is_sof_widget(p->sink)) {
+					ret = sof_route_setup(sdev, widget, p->sink);
+					if (ret < 0)
+						return ret;
+				}
+		}
+	} else {
+		for_each_dapm_widgets(list, i, widget) {
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_source_path(widget, p)
+				if (p->source->dobj.private && is_sof_widget(p->source)) {
+					ret = sof_route_setup(sdev, p->source, widget);
+					if (ret < 0)
+						return ret;
+				}
+		}
+	}
+
+	return 0;
+}
+
+static struct snd_sof_widget *sof_find_pipeline_by_id(struct snd_sof_dev *sdev, int pipeline_id)
+{
+	struct snd_sof_widget *swidget;
+
+	list_for_each_entry(swidget, &sdev->widget_list, list)
+		if (swidget->id == snd_soc_dapm_scheduler && swidget->pipeline_id == pipeline_id)
+			return swidget;
+
+	return NULL;
+}
+
+int sof_pipeline_setup(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget_list *list,
+		       struct snd_sof_pcm *spcm, int dir)
+{
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_widget *w;
+	int i, ret;
+
+	/* set up widgets in the list */
+	for_each_dapm_widgets(list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+		struct snd_sof_widget *pipe_widget, *temp_pipe_widget;
+		bool pipeline_found = false;
+
+		if (!swidget)
+			continue;
+
+		ret = sof_widget_setup(sdev, swidget);
+		if (ret < 0)
+			return ret;
+
+		/*
+		 * find pipeline widget for the pipeline that this widget belongs to and
+		 * add it to the list of pipelines if it doesn't exist already
+		 */
+		pipe_widget = sof_find_pipeline_by_id(sdev, swidget->pipeline_id);
+		if (!pipe_widget)
+			return -EINVAL;
+
+		temp_pipe_widget = kmemdup(pipe_widget, sizeof(*pipe_widget), GFP_KERNEL);
+		if (!temp_pipe_widget)
+			return -ENOMEM;
+
+		list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+			if (w->comp_id == temp_pipe_widget->comp_id) {
+				pipeline_found = true;
+				break;
+			}
+		}
+		if (!pipeline_found)
+			list_add(&temp_pipe_widget->list, &spcm->stream[dir].pipeline_list);
```

Comment:
the list is separae for each stream. So no need to protect. But I dont have to free on the else path because hw_params failure always results in hw_free and the previously allocated temp wdgets will be freed there. I will add a comment.